### PR TITLE
IAsyncEnumerable support in Dataflow

### LINKED ISSUE #30863: TPL Dataflow TransformManyBlock support for IAsyncEnumerable
Currently TransformManyBlock is expected to output an IEnumerable<K>, it would be useful to also have support for an IAsyncEnumerable<K> output

Additional question; why is TPL dataflow in this repository? Seems like an obvious candidate to split into its own repo

# Rationale and Usage

The `TransforManyBlock` of the `System.Threading.Tasks.Dataflow` library currently supports two types of transformation functions:

- `Func<TInput, IEnumerable<TOutput>>`
- `Func<TInput, Task<IEnumerable<TOutput>>>`

.Net Core 3.0 (.Net Standard 2.1) introduced the `IAsyncEnumerable<T>` interface which exposes an enumerator that has a `MoveNextAsync()` method that can be awaited. It is generally presented as a replacement of the `Task<IEnumerable<T>>` type, and has been adopted in a lot of async-related APIs.

In the `TransforManyBlock` context, it could enable some transform functions where we can make asynchronous calls in between yielding results.

For example, if someone uses a sync method inside the transformation function :

```cs
var tb = new TransformManyBlock<T,U>(input =>
{
    var iterator = GetItemQueryIterator<U>(input);
    while (iterator.HasMoreResults)
    {
        foreach(var output in iterator.ReadNext())
        {
            yield return output;
        }
    }
});
```

And wants to migrate it to its async version, he has to modify the code by introducing an instance of an implementation class of the `IEnumerable<T>` interface (here a `List`) that he has to fill in "manually" :

```cs
var tb = new TransformManyBlock<T,U>(async input =>
{
    var iterator = GetItemQueryIterator<U>(input);
    var outputs = new List<U>();
    while (iterator.HasMoreResults)
    {
        foreach(var output in await iterator.ReadNextAsync())
        {
            outputs.Add(output);
        }
    }
    return outputs;
});
```

While he expects to only have to activate async/await mechanisms of the .Net framework :

```cs
var tb = new TransformManyBlock<T,U>(async input =>
{
    var iterator = GetItemQueryIterator<U>(input);
    while (iterator.HasMoreResults)
    {
        foreach(var output in await iterator.ReadNextAsync())
        {
            yield return output;
        }
    }
});
```

# Proposed API

Such a quick and easy migration can be made possible by adding two new constructors for the `TransforManyBlock` class.

```cs
using System.Collections.Generic;

namespace System.Threading.Tasks.Dataflow
{
    public sealed partial class TransformManyBlock<TInput, TOutput>
    {
        public TransformManyBlock(Func<TInput, IAsyncEnumerable<TOutput>> transform) { }
        public TransformManyBlock(Func<TInput, IAsyncEnumerable<TOutput>> transform, ExecutionDataflowBlockOptions dataflowBlockOptions) { }
    }
}
```

@stephentoub mentioned: though as long as we're adding support for `IAsyncEnumerable<T>` to the library, we should probably also add a `ReadAllAsync` extension method, as we did for `System.Threading.Channels`.

```C#
public static class DataflowBlock
{
    public static IAsyncEnumerable<TOutput> ReceiveAllAsync<TOutput>(this ISourceBlock<TOutput> source, CancellationToken cancellationToken = default);
}
```

### PR DESCRIPTION
**Pull request by manandre**
Contributes to #30863

cc @onionhammer @stephentoub

### CODE REVIEW

**Comment by manandre** [comment]
I am not proud of the `ToEnumerable` extension method. Do you see a way to enhance or even remove it?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        {
+            Debug.Assert(transformFunction != null, "Function to invoke is required.");
+
+            bool userDelegateSucceeded = false;
+            try
+            {
+                // Run the user transform and store the results.
+                IAsyncEnumerable<TOutput> outputItems = transformFunction(messageWithId.Key);
+                userDelegateSucceeded = true;
+                ValueTask _ = StoreOutputItemsAsync(messageWithId, outputItems);
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:111]: This is not valid.  The task needs to be awaited.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                // If this exception represents cancellation, swallow it rather than shutting down the block.
+                if (!Common.IsCooperativeCancellation(exc)) throw;
+            }
+            finally
+            {
+                // If the user delegate failed, store an empty set in order
+                // to update the bounding count and reordering buffer.
+                if (!userDelegateSucceeded)
+                {
+                    ValueTask _ = StoreOutputItemsAsync(messageWithId, (IAsyncEnumerable<TOutput>)null);
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:124]: Ditto

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+                    // here itemCopy remains null, so that base.AddItem will finish our interactions with the reordering buffer
+                }
+                else
+                {
+                    // We're not the next item, and we're not trusted, so copy the data into a list.
+                    // We need to enumerate outside of the lock in the base class.
+                    int itemCount = 0;
+                    try
+                    {
+                        itemCopy = item.ToEnumerable().ToList(); // itemCopy will remain null in the case of exception
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:207]: We definitely do not want to do this.  It breaks the whole reason for having an IAsyncEnumerable.  This can just be:
```C#
itemCopy = new List<TOutput>();
await foreach (TOutput element in item.ConfigureAwait(false))
{
    itemCopy.Add(element);
}
```

File: src/libraries/System.Threading.Tasks.Dataflow/src/Internal/AsyncEnumerable.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Internal/AsyncEnumerable.cs:1]: Delete this whole file.  We must not block like this.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 5]
```diff
@@ -1381,6 +1381,34 @@ private object DebuggerDisplayContent
...
+        /// <exception cref="System.ArgumentNullException">The <paramref name="source"/> is null (Nothing in Visual Basic).</exception>
+        public static IAsyncEnumerable<TOutput> ReceiveAllAsync<TOutput>(this ISourceBlock<TOutput> source, CancellationToken cancellationToken = default)
+        {
+            // Validate arguments
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            return source.ReceiveAllAsyncCore(cancellationToken);
+        }
+
+        private static async IAsyncEnumerable<TOutput> ReceiveAllAsyncCore<TOutput>(this ISourceBlock<TOutput> source, [EnumeratorCancellation] CancellationToken cancellationToken)
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs:1402]: You can make this a local function in ReceiveAllAsync.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 6]
```diff
@@ -1381,6 +1381,34 @@ private object DebuggerDisplayContent
             object IDebuggerDisplay.Content { get { return DebuggerDisplayContent; } }
         }
         #endregion
+
+        #region ReceiveAllAsync
+
+#if NETSTANDARD2_1
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs:1387]: This should be separated out into its own partial file.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+{
+    public partial class TransformManyBlock<TInput, TOutput>
+    {
+        /// <summary>Initializes the <see cref="TransformManyBlock{TInput,TOutput}"/> with the specified function.</summary>
+        /// <param name="transform">
+        /// The function to invoke with each data element received.  All of the data from the returned <see cref="System.Collections.Generic.IAsyncEnumerable{TOutput}"/>
+        /// will be made available as output from this <see cref="TransformManyBlock{TInput,TOutput}"/>.
+        /// </param>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="transform"/> is null (Nothing in Visual Basic).</exception>
+        [SuppressMessage("Microsoft.Design", "CA1006:DoNotNestGenericTypesInMemberSignatures")]
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:17]: Why is this needed?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+        /// <summary>Initializes the <see cref="TransformManyBlock{TInput,TOutput}"/> with the specified function and <see cref="ExecutionDataflowBlockOptions"/>.</summary>
+        /// <param name="transform">
+        /// The function to invoke with each data element received.  All of the data from the returned in the <see cref="System.Collections.Generic.IAsyncEnumerable{TOutput}"/>
+        /// will be made available as output from this <see cref="TransformManyBlock{TInput,TOutput}"/>.
+        /// </param>
+        /// <param name="dataflowBlockOptions">The options with which to configure this <see cref="TransformManyBlock{TInput,TOutput}"/>.</param>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="transform"/> is null (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="dataflowBlockOptions"/> is null (Nothing in Visual Basic).</exception>
+        [SuppressMessage("Microsoft.Design", "CA1006:DoNotNestGenericTypesInMemberSignatures")]
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:30]: Why is this needed?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+            {
+                var thisBlock = ((TransformManyBlock<TInput, TOutput>)state!) as IDataflowBlock;
+                Debug.Assert(completed.IsFaulted, "The source must be faulted in order to trigger a target completion.");
+                thisBlock.Fault(completed.Exception!);
+            }, this, CancellationToken.None, Common.GetContinuationOptions() | TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
+
+            // Handle async cancellation requests by declining on the target
+            Common.WireCancellationToComplete(
+                dataflowBlockOptions.CancellationToken, Completion, state => ((TargetCore<TInput>)state!).Complete(exception: null, dropPendingMessages: true), _target);
+#if FEATURE_TRACING
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:89]: This should not be necessary.  It should be set for the 2.1 configuration.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+            DataflowEtwProvider etwLog = DataflowEtwProvider.Log;
+            if (etwLog.IsEnabled())
+            {
+                etwLog.DataflowBlockCreated(this, dataflowBlockOptions);
+            }
+#endif
+        }
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:96]: It's unfortunate how much duplication there is between the existing shared ctor and this one.  I understand you've done so because of the different builds, but it'd be really nice to find a way where the majority of the duplication could be avoided.

**Review by stephentoub — COMMENTED** [review]
Thanks for your interest in working on this.

I did a quick initial pass through and left some comments to be addressed.  I'll do a more in-depth pass subsequently.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs [hunk 11]
```diff
@@ -86,19 +84,18 @@ public TransformManyBlock(Func<TInput, IAsyncEnumerable<TOutput>> transform, Exe
...
                 etwLog.DataflowBlockCreated(this, dataflowBlockOptions);
             }
-#endif
         }
 
         /// <summary>Processes the message with a user-provided transform function that returns an async enumerable.</summary>
         /// <param name="transformFunction">The transform function to use to process the message.</param>
         /// <param name="messageWithId">The message to be processed.</param>
-        private void ProcessMessage(Func<TInput, IAsyncEnumerable<TOutput>> transformFunction, KeyValuePair<TInput, long> messageWithId)
+        private async void ProcessMessage(Func<TInput, IAsyncEnumerable<TOutput>> transformFunction, KeyValuePair<TInput, long> messageWithId)
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:98]: This needs to be awaited as well, or some other mechanism for communicating to the TargetCore that the operations associated with this message have been handled.  Take a look at the existing Task-based async transform to get a sense for what I mean.

File: src/libraries/System.Threading.Tasks.Dataflow/ref/System.Threading.Tasks.Dataflow.cs [hunk 12]
```diff
@@ -131,6 +131,9 @@ public static partial class DataflowBlock
         public static System.Threading.Tasks.Task<bool> OutputAvailableAsync<TOutput>(this System.Threading.Tasks.Dataflow.ISourceBlock<TOutput> source) { throw null; }
         public static System.Threading.Tasks.Task<bool> OutputAvailableAsync<TOutput>(this System.Threading.Tasks.Dataflow.ISourceBlock<TOutput> source, System.Threading.CancellationToken cancellationToken) { throw null; }
         public static bool Post<TInput>(this System.Threading.Tasks.Dataflow.ITargetBlock<TInput> target, TInput item) { throw null; }
+#if NETSTANDARD2_1
+        public static System.Collections.Generic.IAsyncEnumerable<TOutput> ReceiveAllAsync<TOutput>(this System.Threading.Tasks.Dataflow.ISourceBlock<TOutput> source, System.Threading.CancellationToken cancellationToken = default) { throw null; }
```
[the remark this answers is not in the document]
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/ref/System.Threading.Tasks.Dataflow.cs]: Agreed.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 13]
```diff
@@ -1381,6 +1381,34 @@ private object DebuggerDisplayContent
...
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            return source.ReceiveAllAsyncCore(cancellationToken);
+        }
+
+        private static async IAsyncEnumerable<TOutput> ReceiveAllAsyncCore<TOutput>(this ISourceBlock<TOutput> source, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            while (await source.OutputAvailableAsync(cancellationToken).ConfigureAwait(false))
+            {
+                yield return await source.ReceiveAsync(cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: OK.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 1 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs]: `ValueTask` is now awaited, but this leads to an `async void` method... dangerous, no? Is there a better "pattern" to do that?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 2 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs]: Ditto

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 3 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs]: Great! So simple... I like it 😃

File: src/libraries/System.Threading.Tasks.Dataflow/src/Internal/AsyncEnumerable.cs (hunk 4 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Internal/AsyncEnumerable.cs]: Indeed!

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs (hunk 5 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: Done.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs (hunk 6 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: Moved to a dedicated partial file!

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 7 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs]: Removed!
Copy/paste is not always your friend 😋

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 8 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs]: Ditto

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 10 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:93]: I tried multiple approaches to avoid duplication but none of them succeeded 😞
Do you have a specific idea in mind?

Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:93]: 💡
Can we build the Dataflow component only for `$(NetCoreAppCurrent)` (currently `net5.0`)?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 11 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:98]: I have added calls to `_target.Complete` as done in existing Task-based async transform.
Is it enough?

**Comment by manandre** [comment]
@stephentoub I have also reworked the test part but there is a test failure on Helix that I do not manage to analyze nor reproduce locally. Can you please help me?

**Comment by tarekgh** [comment]
@manandre it looks this PR parked for long time. could you please force the CI rerun on it again just to run against the latest? you can do that by adding any trivial commit to it. Also, what was pending on this change? 

@stephentoub do you have any more comments on this PR?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs (hunk 10 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs:35]: I found a possible solution by introducing a shared Initialize method to replace the private constructor.
Better ?

**Comment by manandre** [comment]
I have rebased on master branch but the build still fails on a unit test failure that I managed to reproduce locally (by running it in loop until failure):
```
E:\git\runtime\artifacts\bin\System.Threading.Tasks.Dataflow.Tests\net6.0-Debug>E:\git\runtime\artifacts\bin\testhost\net6.0-windows-Debug-x86\dotnet exec --runtimeconfig System.Threading.Tasks.Dataflow.Tests.runtimeconfig.json --depsfile System.Threading.Tasks.Dataflow.Tests.deps.json xunit.console.dll System.Threading.Tasks.Dataflow.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing -method System.Threading.Tasks.Dataflow.Tests.DataflowBlockExtensionTests.ReceiveAllAsync*
  Discovering: System.Threading.Tasks.Dataflow.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Threading.Tasks.Dataflow.Tests (found 11 of 335 test cases)
  Starting:    System.Threading.Tasks.Dataflow.Tests (parallel test collections = on, max threads = 8)
    System.Threading.Tasks.Dataflow.Tests.DataflowBlockExtensionTests.ReceiveAllAsync_MultipleEnumerationsToEnd [FAIL]
      System.InvalidOperationException : An attempt was made to transition a task to a final state when it had already completed.
      Stack Trace:
        E:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncTaskMethodBuilderT.cs(443,0): at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(Task`1 task, TResult result)
        E:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncIteratorMethodBuilder.cs(62,0): at System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.Complete()
        E:\git\runtime\src\libraries\System.Threading.Tasks.Dataflow\src\Base\DataflowBlock.IAsyncEnumerable.cs(33,0): at System.Threading.Tasks.Dataflow.DataflowBlock.<ReceiveAllAsync>g__ReceiveAllAsyncCore|43_0[TOutput](IReceivableSourceBlock`1 source, CancellationToken cancellationToken)+MoveNext()
        E:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs(42,0): at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
           at System.Threading.Tasks.Dataflow.DataflowBlock.<ReceiveAllAsync>g__ReceiveAllAsyncCore|43_0[TOutput](IReceivableSourceBlock`1 source, CancellationToken cancellationToken)+System.Collections.Generic.IAsyncEnumerator<TOutput>.MoveNextAsync()
        E:\git\runtime\src\libraries\System.Threading.Tasks.Dataflow\tests\Dataflow\DataflowBlockExtensionTests.IAsyncEnumerable.cs(166,0): at System.Threading.Tasks.Dataflow.Tests.DataflowBlockExtensionTests.ReceiveAllAsync_MultipleEnumerationsToEnd()
        --- End of stack trace from previous location ---
  Finished:    System.Threading.Tasks.Dataflow.Tests
=== TEST EXECUTION SUMMARY ===
   System.Threading.Tasks.Dataflow.Tests  Total: 19, Errors: 0, Failed: 1, Skipped: 0, Time: 0,787s
```
Unfortunately, I do not understand what happens here. I would need some help.

**Comment by tarekgh** [comment]
@stephentoub could you please help @manandre with his question in https://github.com/dotnet/runtime/pull/37876#issuecomment-735279502?

**Comment by stephentoub** [comment]
> I have rebased on master branch but the build still fails on a unit test failure that I managed to reproduce locally (by running it in loop until failure)
> Unfortunately, I do not understand what happens here. I would need some help.

That error suggests that something is causing the continuations in the async iterator to be executed multiple times, likely due to a bug in the implementation of the blocks as part of the changes made in this PR.

In the name of moving this PR forward, I suggest scoping it back to the simpler ReceiveAllAsync and project updates, rather than also supporting the much more complicated and problematic new constructors.

**Comment by manandre** [comment]
@stephentoub Unfortunately, these random unit test failures only occur with the _ReceiveAllAsync_* tests, which rely on the already existing `BufferBlock` class (and not modified by this PR).

File: src/libraries/System.Threading.Tasks.Dataflow/tests/Dataflow/DataflowBlockExtensionTests.IAsyncEnumerable.cs [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+            var source = new BufferBlock<int>();
+            IAsyncEnumerator<int> e = source.ReceiveAllAsync().GetAsyncEnumerator();
+
+            ValueTask<bool> vt;
+            if (completeWhilePending)
+            {
+                source.Complete();
+                vt = e.MoveNextAsync();
+                Assert.False(await vt);
+                Assert.True(vt.IsCompletedSuccessfully);
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/tests/Dataflow/DataflowBlockExtensionTests.IAsyncEnumerable.cs:41]: Operations like this are not safe.  You can't touch `vt` after you've awaited it.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.IAsyncEnumerable.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+            // Validate arguments
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            return ReceiveAllAsyncCore(source, cancellationToken);
+
+            static async IAsyncEnumerable<TOutput> ReceiveAllAsyncCore(IReceivableSourceBlock<TOutput> source, [EnumeratorCancellation] CancellationToken cancellationToken)
+            {
+                while (await source.OutputAvailableAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    if (source.TryReceive(out TOutput item))
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.IAsyncEnumerable.cs:28]:
```suggestion
                    while (source.TryReceive(out TOutput item))
```

**Comment by stephentoub** [comment]
> Unfortunately, these random unit test failures only occur with the ReceiveAllAsync* tests, which rely on the already existing BufferBlock class (and not modified by this PR).

Can you narrow the PR down to just ReceiveAllAsync, its tests, and the minimal project changes necessary to add the netstandard2.1 configuration? (i.e. leave out the rest of the TransformXx block tests).  It's harder to prove that nothing else relevant has changed when there's so much other stuff here.

If it still repros after that, it should be easier to diagnose.

Thanks.

File: src/libraries/System.Threading.Tasks.Dataflow/tests/Dataflow/DataflowBlockExtensionTests.IAsyncEnumerable.cs (hunk 14 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/tests/Dataflow/DataflowBlockExtensionTests.IAsyncEnumerable.cs]: OK. I removed this part of the test.

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.IAsyncEnumerable.cs (hunk 15 above)
Comment (manandre) [reply src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.IAsyncEnumerable.cs]: Agreed!

**Comment by manandre** [comment]
PR narrowed down to `ReceiveAllAsync` support. The random issue is still present. Any clue on how to diagnose such an issue?

**Comment by stephentoub** [comment]
@eerhardt, how should the project changes here interact with the project changes you made to dataflow?

**Comment by eerhardt** [comment]
> how should the project changes here interact with the project changes you made to dataflow?

They are conflicting. I also added a `$(NetCoreAppCurrent)` build to this project in #48667.

File: src/libraries/System.Threading.Tasks.Dataflow/ref/System.Threading.Tasks.Dataflow.csproj [hunk 16]
```diff
@@ -1,9 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard1.0;netstandard1.1;net461</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.1;netstandard2.0;netstandard1.0;netstandard1.1;net461</TargetFrameworks>
```
Comment (eerhardt) [inline src/libraries/System.Threading.Tasks.Dataflow/ref/System.Threading.Tasks.Dataflow.csproj:3]: I don't believe we need to add a `$(NetCoreAppCurrent)` build for the ref assembly. Just adding `netstandard2.1` should be enough.

**Comment by stephentoub** [comment]
> I also added a $(NetCoreAppCurrent) build to this project in #48667.

Yes, but doesn't this exclude it from the package?
https://github.com/dotnet/runtime/pull/48667/files#diff-4e3779396c883b3f673a3665a87b1f280c1cc0c65f5a441042c3ac232d13455bR4

**Comment by eerhardt** [comment]
> So the correct change for this PR is just to add the netstandard2.1 ref assembly?

I believe the correct change for this PR is to add the `netstandard2.1` build for both the ref and the implementation assembly.

You want it in the impl assembly as well, because that goes into the NuGet package.

**Comment by stephentoub** [comment]
> You want it in the impl assembly as well, because that goes into the NuGet package.

Ah, of course.

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks!
